src/vsr/sync.zig

```diff
@@ -0,0 +1,288 @@
+const std = @import("std");
+const assert = std.debug.assert;
+const maybe = stdx.maybe;
+
+const vsr = @import("../vsr.zig");
+const constants = @import("../constants.zig");
+const stdx = @import("../stdx.zig");
+
+pub const SyncStage = union(enum) {
+    /// Not syncing.
+    none,
+
+    /// The commit lifecycle is in a stage that cannot be interrupted/cancelled.
+    /// We are waiting until that stage completes.
+    /// When it completes, we will terminate the commit and resume sync.
+    cancel_commit,
+
+    /// Waiting for `Grid.cancel()`.
+    cancel_grid,
+
+    /// We need to sync, but are waiting for a usable `sync_target_max`.
+    request_target,
+
+    request_trailers: struct {
+        target: SyncTarget,
+        manifest: SyncTrailer = .{},
+        free_set: SyncTrailer = .{},
+        client_sessions: SyncTrailer = .{},
+        /// Our new VSRState.previous_checkpoint_id.
+        /// Arrives with command=sync_free_set.
+        previous_checkpoint_id: ?u128 = null,
+        /// The checksum of the prepare corresponding to checkpoint_op.
+        /// Arrives with command=sync_client_sessions.
+        checkpoint_op_checksum: ?u128 = null,
+
+        pub fn done(self: *const @This()) bool {
+            if (self.free_set.done) assert(self.previous_checkpoint_id != null);
+            if (self.client_sessions.done) assert(self.checkpoint_op_checksum != null);
+
+            return self.manifest.done and self.free_set.done and self.client_sessions.done;
+        }
+    },
+
+    superblock_update: struct {
+        target: SyncTarget,
+        previous_checkpoint_id: u128,
+        checkpoint_op_checksum: u128,
+    },
+
+    done,
+
+    pub fn valid_transition(from: std.meta.Tag(SyncStage), to: std.meta.Tag(SyncStage)) bool {
+        return switch (from) {
+            .none => to == .cancel_commit or
+                to == .cancel_grid or
+                to == .request_target,
+            .cancel_commit => to == .cancel_grid,
+            .cancel_grid => to == .request_target,
+            .request_target => to == .request_target or
+                to == .request_trailers,
+            .request_trailers => to == .request_trailers or
+                to == .superblock_update,
+            .superblock_update => to == .request_trailers or
+                to == .done,
+            .done => to == .request_trailers or
+                to == .none,
+        };
+    }
+
+    pub fn target(stage: *const SyncStage) ?SyncTarget {
+        return switch (stage.*) {
+            .none,
+            .cancel_commit,
+            .cancel_grid,
+            .request_target,
+            .done,
+            => null,
+            .request_trailers => |s| s.target,
+            .superblock_update => |s| s.target,
+        };
+    }
+};
+
+pub const SyncTargetCandidate = struct {
```

Comment:
Would be good to motivate via comment why we duplicate the `SyncTarget` type here.

Reply:
Is this risk enough that we need to, or could we thin the code to only have `SyncTarget`?

Reply:
I added a comment explaining. I think it is worth keeping them separate -- mixing up a canonical and candidate target is potentially dangerous. (Though it will become less dangerous once we include the checkpoint id in prepares).